+nosnoc/+model/Objects.m

```diff
@@ -0,0 +1,421 @@
+
+            % Lambda and gap functions
+            lambda = SX.sym(['lambda_' ball1.name '_' ball2.name]);
+            c = sum((ball1.x - ball2.x).^2) - (ball1.r + ball2.r)^2;
+
+            % Get normals
+            nabla_c_x1 = c.jacobian(ball1.x)';
+            nabla_c_x2 = c.jacobian(ball2.x)';
+
+            % Update dynamics of both balls
+            ball1.x_dot = ball1.x_dot + nabla_c_x1*lambda;
+            ball2.x_dot = ball2.x_dot + nabla_c_x2*lambda;
+
+            normal = [nabla_c_x1;nabla_c_x2];
+            obj.normal = [obj.normal;normal];
+            obj.normal0 = [obj.normal0;0;1;0;-1];
+
+            if mu_f
+                if ball1.n_dim == 2
+                    tangent = SX.sym('tangent', 2);
+                    obj.tangent = [obj.tangent; tangent];
+                    obj.tangent0 = [obj.tangent0;1;0];
+                    v_t = SX.sym('v_t', 2);
+                    obj.v_t = [obj.v_t; v_t];
+                    lambda_t = SX.sym('lambda_t', 2);
+                    obj.lambda_t = [obj.lambda_t; lambda_t];
+                    gamma_f = SX.sym('gamma_f', 1);
+                    obj.gamma_f = [obj.gamma_f; gamma_f];
+                    normal_lift = SX.sym('normal_lift', 4);
+                    obj.normal_lift = [obj.normal_lift; normal_lift];
+
+                    ball1.x_dot = ball1.x_dot + lambda_t(1)*tangent - lambda_t(2)*tangent;
+                    ball2.x_dot = ball2.x_dot - lambda_t(1)*tangent + lambda_t(2)*tangent;
+                    
+                    g_friction = [normal_lift - normal; % Lift normal (helps with convergence) TODO(@anton) allow for not lifting
+                        tangent - [-normal_lift(2);normal_lift(1)]; % get tangent (lifted again) TODO(@anton) allow for not lifting
+                        v_t(1) - (dot(ball1.x_dot_lift, tangent)/dot(tangent,tangent)); % Tangent velocities of both objects 
+                        v_t(2) - (dot(ball2.x_dot_lift, tangent)/dot(tangent,tangent))];
+                    obj.g_friction = [obj.g_friction; g_friction];
+                    
+                    G_friction = [v_t(1) - v_t(2) + gamma_f;
+                        v_t(2) - v_t(1) + gamma_f;
+                        mu_f*lambda - lambda_t(1) - lambda_t(2);
+                        lambda_t(1)]; % TODO(@anton) can we avoid this as it violates MPCC-LICQ
+                    obj.G_friction = [obj.G_friction; G_friction];
+                    H_friction = [lambda_t; gamma_f;lambda_t(2)];
+                    obj.H_friction = [obj.H_friction; H_friction];
+                    
+                else
+                    error("3d friction not yet supported")
+                end
+            end
+            
+            % update model
+            obj.c = vertcat(obj.c,c);
+            obj.lambda = vertcat(obj.lambda, lambda);
+        end
+
+        function addEllipseBall(obj, ellipse, ball, mu_f)
+            import casadi.*
+            n_dim = ball.n_dim;
+
```

Comment:
every function would benefit greatly of a 1-2 line descripiton (most names are self explaining, slightly more details would be welcome